File: fraiseql-swift/Tests/FraiseQLTests/ExportTypesTests.swift
import XCTest
import Foundation
@testable import FraiseQL

final class ExportTypesTests: XCTestCase {
  override func setUp() {
    super.setUp()
    Schema.reset()
  }

  override func tearDown() {
    super.tearDown()
    Schema.reset()
  }

  func testExportTypesMinimalSingleType() throws {
    Schema.registerType("User", fields: [
      "id": ["type": "ID", "nullable": false],
      "name": ["type": "String", "nullable": false],
      "email": ["type": "String", "nullable": false],
    ], description: "User in the system")

    let json = Schema.exportTypes(pretty: true)
    let data = json.data(using: .utf8)!
    let parsed = try JSONSerialization.jsonObject(with: data) as! [String: Any]

    XCTAssertNotNil(parsed["types"])
    let types = parsed["types"] as! [[String: Any]]
    XCTAssertEqual(types.count, 1)

    XCTAssertNil(parsed["queries"])
    XCTAssertNil(parsed["mutations"])

    let userDef = types[0]
    XCTAssertEqual(userDef["name"] as? String, "User")
    XCTAssertEqual(userDef["description"] as? String, "User in the system")
  }

  func testExportTypesMultipleTypes() throws {
    Schema.registerType("User", fields: [
      "id": ["type": "ID", "nullable": false],
    ])
    Schema.registerType("Post", fields: [
      "id": ["type": "ID", "nullable": false],
      "title": ["type": "String", "nullable": false],
    ])

    let json = Schema.exportTypes(pretty: true)
    let data = json.data(using: .utf8)!
    let parsed = try JSONSerialization.jsonObject(with: data) as! [String: Any]

    let types = parsed["types"] as! [[String: Any]]
    XCTAssertEqual(types.count, 2)

    let typeNames = types.compactMap { $0["name"] as? String }
    XCTAssert(typeNames.contains("User"))
    XCTAssert(typeNames.contains("Post"))
  }

  func testExportTypesNoQueries() throws {
    Schema.registerType("User", fields: [
      "id": ["type": "ID", "nullable": false],
    ])

    let json = Schema.exportTypes(pretty: true)
    let data = json.data(using: .utf8)!
    let parsed = try JSONSerialization.jsonObject(with: data) as! [String: Any]

    XCTAssertNotNil(parsed["types"])
    XCTAssertNil(parsed["queries"])
    XCTAssertNil(parsed["mutations"])
  }

  func testExportTypesCompactFormat() throws {
    Schema.registerType("User", fields: [
      "id": ["type": "ID", "nullable": false],
    ])

    let compact = Schema.exportTypes(false)
    let pretty = Schema.exportTypes(true)

    // Both should be valid JSON
    let compactData = compact.data(using: .utf8)!
    let prettyData = pretty.data(using: .utf8)!
    XCTAssertNoThrow(try JSONSerialization.jsonObject(with: compactData))
    XCTAssertNoThrow(try JSONSerialization.jsonObject(with: prettyData))

    // Compact should be smaller
    XCTAssertLessThanOrEqual(compact.count, pretty.count)
  }

  func testExportTypesPrettyFormat() throws {
    Schema.registerType("User", fields: [
      "id": ["type": "ID", "nullable": false],
    ])

    let json = Schema.exportTypes(true)
    XCTAssert(json.contains("\n"))

    let data = json.data(using: .utf8)!
    XCTAssertNoThrow(try JSONSerialization.jsonObject(with: data))
  }

  func testExportTypesFile() throws {
    Schema.registerType("User", fields: [
      "id": ["type": "ID", "nullable": false],
      "name": ["type": "String", "nullable": false],
    ])

    let tmpFile = "/tmp/fraiseql_types_test_swift.json"
    try? FileManager.default.removeItem(atPath: tmpFile)

    Schema.exportTypesFile(tmpFile)

    XCTAssert(FileManager.default.fileExists(atPath: tmpFile))

    let content = try String(contentsOfFile: tmpFile, encoding: .utf8)
    let data = content.data(using: .utf8)!
    let parsed = try JSONSerialization.jsonObject(with: data) as! [String: Any]

    XCTAssertNotNil(parsed["types"])
    let types = parsed["types"] as! [[String: Any]]
    XCTAssertEqual(types.count, 1)

    try? FileManager.default.removeItem(atPath: tmpFile)
  }

  func testExportTypesEmpty() throws {
    let json = Schema.exportTypes(true)
    let data = json.data(using: .utf8)!
    let parsed = try JSONSerialization.jsonObject(with: data) as! [String: Any]

    XCTAssertNotNil(parsed["types"])
    let types = parsed["types"] as! [[String: Any]]
    XCTAssertEqual(types.count, 0)
  }
}

File: sdks/community/fraiseql-csharp/tests/FraiseQL.Tests/ExportTypesTests.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using Xunit;
using FraiseQL;

namespace FraiseQL.Tests
{
    public class ExportTypesTests : IDisposable
    {
        public ExportTypesTests()
        {
            Schema.Reset();
        }

        public void Dispose()
        {
            Schema.Reset();
        }

        [Fact]
        public void TestExportTypesMinimalSingleType()
        {
            Schema.RegisterType("User", new Dictionary<string, object>
            {
                { "id", new Dictionary<string, object> { { "type", "ID" }, { "nullable", false } } },
                { "name", new Dictionary<string, object> { { "type", "String" }, { "nullable", false } } },
                { "email", new Dictionary<string, object> { { "type", "String" }, { "nullable", false } } }
            }, "User in the system");

            var json = Schema.ExportTypes(true);
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            Assert.True(root.TryGetProperty("types", out var types));
            Assert.Equal(JsonValueKind.Array, types.ValueKind);
            Assert.Equal(1, types.GetArrayLength());

            Assert.False(root.TryGetProperty("queries", out _));
            Assert.False(root.TryGetProperty("mutations", out _));
            Assert.False(root.TryGetProperty("observers", out _));

            var userDef = types[0];
            Assert.Equal("User", userDef.GetProperty("name").GetString());
            Assert.Equal("User in the system", userDef.GetProperty("description").GetString());
        }

        [Fact]
        public void TestExportTypesMultipleTypes()
        {
            Schema.RegisterType("User", new Dictionary<string, object>
            {
                { "id", new Dictionary<string, object> { { "type", "ID" }, { "nullable", false } } },
                { "name", new Dictionary<string, object> { { "type", "String" }, { "nullable", false } } }
            });

            Schema.RegisterType("Post", new Dictionary<string, object>
            {
                { "id", new Dictionary<string, object> { { "type", "ID" }, { "nullable", false } } },
                { "title", new Dictionary<string, object> { { "type", "String" }, { "nullable", false } } }
            });

            var json = Schema.ExportTypes(true);
            using var doc = JsonDocument.Parse(json);
            var types = doc.RootElement.GetProperty("types");

            Assert.Equal(2, types.GetArrayLength());

            var typeNames = types.EnumerateArray().Select(t => t.GetProperty("name").GetString()).ToList();
            Assert.Contains("User", typeNames);
            Assert.Contains("Post", typeNames);
        }

        [Fact]
        public void TestExportTypesNoQueries()
        {
            Schema.RegisterType("User", new Dictionary<string, object>
            {
                { "id", new Dictionary<string, object> { { "type", "ID" }, { "nullable", false } } }
            });

            var json = Schema.ExportTypes(true);
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            Assert.True(root.TryGetProperty("types", out _));
            Assert.False(root.TryGetProperty("queries", out _));
            Assert.False(root.TryGetProperty("mutations", out _));
        }

        [Fact]
        public void TestExportTypesCompactFormat()
        {
            Schema.RegisterType("User", new Dictionary<string, object>
            {
                { "id", new Dictionary<string, object> { { "type", "ID" }, { "nullable", false } } }
            });

            var compact = Schema.ExportTypes(false);
            var pretty = Schema.ExportTypes(true);

            Assert.True(compact.Length <= pretty.Length);

            var compactDoc = JsonDocument.Parse(compact);
            Assert.True(compactDoc.RootElement.TryGetProperty("types", out _));

            var prettyDoc = JsonDocument.Parse(pretty);
            Assert.True(prettyDoc.RootElement.TryGetProperty("types", out _));
        }

        [Fact]
        public void TestExportTypesPrettyFormat()
        {
            Schema.RegisterType("User", new Dictionary<string, object>
            {
                { "id", new Dictionary<string, object> { { "type", "ID" }, { "nullable", false } } }
            });

            var json = Schema.ExportTypes(true);
            Assert.Contains("\n", json);

            using var doc = JsonDocument.Parse(json);
            Assert.True(doc.RootElement.TryGetProperty("types", out _));
        }

        [Fact]
        public void TestExportTypesFile()
        {
            Schema.RegisterType("User", new Dictionary<string, object>
            {
                { "id", new Dictionary<string, object> { { "type", "ID" }, { "nullable", false } } },
                { "name", new Dictionary<string, object> { { "type", "String" }, { "nullable", false } } }
            });

            var tmpFile = "/tmp/fraiseql_types_test_csharp.json";
            if (File.Exists(tmpFile))
                File.Delete(tmpFile);

            Schema.ExportTypesFile(tmpFile);

            Assert.True(File.Exists(tmpFile));

            var content = File.ReadAllText(tmpFile);
            using var doc = JsonDocument.Parse(content);
            var types = doc.RootElement.GetProperty("types");
            Assert.Equal(1, types.GetArrayLength());

            File.Delete(tmpFile);
        }

        [Fact]
        public void TestExportTypesEmpty()
        {
            var json = Schema.ExportTypes(true);
            using var doc = JsonDocument.Parse(json);
            var types = doc.RootElement.GetProperty("types");

            Assert.Equal(JsonValueKind.Array, types.ValueKind);
            Assert.Equal(0, types.GetArrayLength());
        }
    }
}
